### PR TITLE
add AQL bit handling functions

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

This PR add the following bit handling functionality to AQL:

- BIT_AND(array): and-combined result
- BIT_OR(array): or-combined result
- BIT_XOR(array): xor-combined result
- BIT_NEGATE(value, bits): bitwise negation of `value`, with a mask of bits` length
- BIT_TEST(value, index): test if bit at position `index` is set in `value` (indexes are 0-based)
- BIT_POPCOUNT(value): return number of bits set in `value`
- BIT_SHIFT_LEFT(value, shift, bits): bitwise shift-left of `value` by `shift` bits, with a mask of `bits` length
- BIT_SHIFT_RIGHT(value, shift, bits): bitwise shift-right of `value` by `shift` bits, with a mask of `bits` length
- BIT_CONSTRUCT(array): construct a number with bits set at the positions given in the array
- BIT_DECONSTRUCT(value): deconstruct a number into an array with the set bits
- BIT_TO_STRING(value): create a bitstring representation from numeric `value`
- BIT_FROM_STRING(value): parse a bitstring representation into a number

`BIT_AND`, `BIT_OR` and `BIT_XOR` are also available as aggregate functions for usage inside COLLECT AGGREGATE.

All above bit operations support unsigned integer values with up to 32 bits. Using values outside the supported range will make any of these bit functions return `null` and register a warning.

The PR also adds binary (base 2) and hexadecimal (base 16) integer literals to AQL. 
These literals can be used where regular (base 10) integer literal can used. 
The prefix for binary integer literals is `0b`, e.g. `0b10101110`.
The prefix for hexadecimal integer literals i `0x`, e.g. `0xabcdef02`.

Binary and hexadecimal integer literals can only be used for unsigned integers.
The maximum supported value is `(2 ^ 32) - 1`, i.e. `0xffffffff` (hexadecimal) or `0b11111111111111111111111111111111` (binary).

This is in RFC phase now, when we have agreed on functionality and behavior, we can backport it to 3.7.

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [x] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [x] :book: CHANGELOG entry made

#### Backports:

- [x] Backports required for: 3.7 

#### Related Information

- [x] Docs PR: https://github.com/arangodb/docs/pull/607

### Testing & Verification

- [x] The behavior in this PR was *manually tested*
- [x] This PR adds tests that were used to verify all changes:
  - [x] Added new C++ **Unit tests**

Link to Jenkins PR run:
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/13483/

### CODE REVIEW

**Review by fceller — APPROVED** [review]
LGTM

**Comment by jsteemann** [comment]
Tests blue

File: CHANGELOG [hunk 1]
```diff
@@ -3,6 +3,51 @@ devel
...
 * Added AQL bit handling functions BIT_AND, BIT_OR, BIT_XOR and BIT_COUNT.
 
+* Fixed some situations of
+  [...]
+  SUBQUERY
+  FILTER
+  LIMIT
+  [...]
+  in AQL queries, yielding incorrect responses. A distributed
+  state within the subquery was not resetted correctly.
```
Comment (Simran-B) [inline CHANGELOG]:
```suggestion
  state within the subquery was not reset correctly.
```

File: CHANGELOG [hunk 2]
```diff
@@ -3,6 +3,51 @@ devel
...
+      aggregator is now removed.
+    - All queries of the form `COLLECT AGGREGATE x = LENGTH(_)` are now executed
+      using the count executor, which can result in significantly improved
+      performance.
+
+* Minor and rare AQL performance improvement, in nested subqueries:
+  LET sq1 ([..] FILTER false == true LET sq2 = (<X>) [..])
+  where sq1 produces no data (e.g. by the above filter) for sq2,
+  the part <X> have been asked two times (second returns empty result),
+  instead of one, if and only if the mainquery executes sq1 exactly one time.
```
Comment (Simran-B) [inline CHANGELOG]:
```suggestion
  instead of one, if and only if the main query executes sq1 exactly one time.
```

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/13612/

File: CHANGELOG (hunk 1 above)
Comment (jsteemann) [reply CHANGELOG]: Unrelated to the changes in this PR, but I fixed it in here as well.

File: CHANGELOG (hunk 2 above)
Comment (jsteemann) [reply CHANGELOG]: Unrelated to the changes in this PR, but I fixed it in here as well.

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/13614/

**Comment by jsteemann** [comment]
Tests blue

**Review by goedderz — COMMENTED** [review]
All parsing functions are implemented s.t. leading zeroes are ignored, even if the size would be larger than 32bits in total. It's a probably unimportant detail without much impact, but does it make more sense than restricting the `0b` literals and strings to 32 chars, and the `0x` literals and strings to 8, even if they have leading zeroes?

File: arangod/Aql/Aggregator.cpp [hunk 3]
```diff
@@ -834,98 +832,179 @@ struct AggregatorCountDistinctStep2 final : public AggregatorCountDistinct {
...
+struct BitFunctionOr {
+  int64_t compute(int64_t value1, int64_t value2) noexcept {
+    return value1 | value2;
+  }
+};
+
+struct BitFunctionXOr {
+  int64_t compute(int64_t value1, int64_t value2) noexcept {
+    return value1 ^ value2;
+  }
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/Aql/Aggregator.cpp]: Changed it all to uint64_t now.

File: arangod/Aql/Functions.cpp [hunk 4]
```diff
@@ -6900,6 +6900,357 @@ AqlValue Functions::Pi(ExpressionContext*, AstNode const&,
...
+/// @brief function BIT_XOR
+AqlValue Functions::BitXOr(ExpressionContext* expressionContext, AstNode const& node,
+                          VPackFunctionParameters const& parameters) {
+  return handleBitOperation(expressionContext, node, parameters, [](uint64_t value1, uint64_t value2) {
+      return value1 ^ value2;
+  });
+}
+
+template<typename T1, typename T2>
+std::tuple<T1, T2, bool> binaryBitFunctionParameters(VPackFunctionParameters const& parameters) {
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/Aql/Functions.cpp]: Moved it to `std::optional` now.

File: arangod/Aql/Functions.cpp [hunk 5]
```diff
@@ -6900,6 +6900,357 @@ AqlValue Functions::Pi(ExpressionContext*, AstNode const&,
...
+                                VPackFunctionParameters const& parameters) {
+  AqlValue const& value = extractFunctionParameterValue(parameters, 0);
+
+  if (value.isNumber()) {
+    VPackSlice v = value.slice();
+    auto [x, valid] = bitOperationValue<uint64_t>(v);
+    if (valid) {
+      // the following code fragment is from https://en.wikipedia.org/wiki/Hamming_weight#Efficient_implementation
+      // TODO: we could optimize this to use hardware-accelerated popcnt if it exists, or gcc's __builtin_popcntll
+      // or such.
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/Aql/Functions.cpp]: Added such code comment.

File: arangod/Aql/Functions.cpp [hunk 6]
```diff
@@ -6900,6 +6900,357 @@ AqlValue Functions::Pi(ExpressionContext*, AstNode const&,
...
+      transaction::Methods* trx = &expressionContext->trx();
+      transaction::BuilderLeaser builder(trx);
+       builder->openArray();
+       uint64_t compare = 1;
+       for (uint64_t i = 0; i < Functions::bitFunctionsMaxSupportedBits; ++i) {
+         if ((num & compare) != 0) {
+           builder->add(VPackValue(i));
+         }
+         compare <<= 1;
+       }
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/Aql/Functions.cpp]: Added such comment, but didn't change the implementation.

File: arangod/Aql/Functions.cpp [hunk 7]
```diff
@@ -6900,6 +6900,357 @@ AqlValue Functions::Pi(ExpressionContext*, AstNode const&,
...
+  auto [testee, index, valid] = binaryBitFunctionParameters<uint64_t, uint64_t>(parameters);
+  if (valid) {
+    TRI_ASSERT(index <= Functions::bitFunctionsMaxSupportedBits);
+
+    char buffer[Functions::bitFunctionsMaxSupportedBits];
+    char* p = &buffer[0];
+
+    if (index > 0) {
+      uint64_t compare = uint64_t(1) << (index - 1);
+      while (compare > 0 && index > 0) {
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/Aql/Functions.cpp]: True. Adjusted it.

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/13624/

File: arangod/Aql/Functions.cpp [hunk 8]
```diff
@@ -6900,6 +6900,357 @@ AqlValue Functions::Pi(ExpressionContext*, AstNode const&,
   return ::numberValue(std::acos(-1.0), true);
 }
 
+template<typename T> 
+std::pair<T, bool> bitOperationValue(VPackSlice input) {
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/Aql/Functions.cpp]: done

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/13627/

**Comment by jsteemann** [comment]
Tests blue

**Review by goedderz — APPROVED** [review]
LGTM

**Pull request merged**